### PR TITLE
fix(Core/Events) Implement Fireworks Spectacular

### LINKED ISSUE #16677: Engineer's Explosive Extravaganza (Midsummer Fireworks Spectacular) event is unimplemented.
### Current Behaviour

Currently there is some [debris](https://github-production-user-asset-6210df.s3.amazonaws.com/38561765/248529029-727b26d3-7e81-4789-8644-700d4bbcce4d.png) in front of the Exodar's entrance as a part of the Midsummer event.

### Expected Blizzlike Behaviour

Timing is weird of course, but around the 4th of July (May have been a different date or nonexistent for non-US realms) a bunch of NPCs/gameobjects should spawn around various capitals. Many reveler NPCs, fireworks vendors, tables, kegs, goblets, and mugs. The NPCs should say various things depending on which area they're attending (and depending on the status of the fireworks show), and the goblets (I believe) should only be spawned for a bit, also dependent on the fireworks show. Every hour from the start of the event (Intuition says no fireworks at the start, every hour following including the last the event should take place) there should be a fireworks show at every capital city as well as Booty Bay in a specific spot. (Gates of Org, Exodar, SW, Ironforge, Silvermoon, UC; close to the flight post in TB, just off the coast of Rut'theran Village, above one of the small paths in Shattrath's upper city, and in the middle of the bay in Booty Bay) This show should go on for about ten minutes with the revelers in the area turning towards it at certain points, and (iirc) cheering at the end.

### Source

Wrath Classic, Wowhead run-downs

### Steps to reproduce the problem

1. Exist on the 4th of July.
2. Observe AzerothCore does not celebrate the United States of America's day of independence.

### Extra Notes

~~The Exodar debris section of this report was brought to you in part by: https://github.com/chromiecraft/chromiecraft/issues/5832.~~ fixed

I have a lot of this particular event sniffed, though very possibly not all. I was sniffing at the top of every hour at every location mentioned in the report as well as Ratchet and Dalaran, neither of which I saw any sort of fireworks display or commotion at, so I believe these two do not have anything of interest related to this event.

### AC rev. hash/commit

dc071dab4cc8edaafbbefd5f765b02d4dcb7613b

### Operating system

Windows 10 x64

### Custom changes or Modules

[mod-ptr-template](https://github.com/heyitsbench/mod-ptr-template)

### PR DESCRIPTION
**Pull request by sudlud**
## Changes Proposed:

This PR proposes changes to:
-  [x] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [x] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes #16677

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [x] Sniffs (remember to share them with the open source community!)
- [x] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

Thanks @heyitsbench for sniffing all firework shows!

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

event runs for 18 hours
.event start 72

firework shows start at the full hour and last for about 11-12 minutes
at the locations there are kegs, tables, mugs and revelers
revelers emote regularly 
revelers cheer at show end
toasting goblets spawn at show end for 10 minutes

Booty Bay Bruisers are replaced by sleeping Drunken Bruisers

visit firework vendors
Fizzlebang Booms
.go c 12523
Ziggle Sparks
.go c 12524
Wizbang Booms
.go c 12525
Lillian Singh
.go c 12526
Jaxxil Sparks
.go c 12527
Kizzak Sparks
.go c 12528

visit firework locations
Stormwind
.go c 420011
Thunder Bluff
.go c 420024
Booty Bay
.go c 420041
Teldrassil
.go c 420122
Ironforge
.go c 420141
Undercity
.go c 420165
Orgrimmar
.go c 420184
Shattrath (Scryer)
.go c 420203
Shattrath (Aldor)
.go c 420213
Exodar
.go c 420221
Silvermoon
.go c 420240

enjoy

## Known Issues and TODO List:

- [ ] reveler gossip texts missing, they're showing Lunar (?) gossips right now, as they're also present during that event
- [ ] reveler orientations need to be updated (manually) - sniffs CO seem to have shown orientations right in the middle of some RP most of the time, could be grepped from sniff reset-orientation-home-messages maybe
- [ ] reveler RP might not be final - they seem to orient themselfes to the firework and back sometimes, also the emotes during could be added to the show script (maybe) to get it right
- [ ] Thunder Bluff Reveler spawns might need a check - they seem to be a bit much, they probably should have much less spawns and waypoint movement

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Kitzunu** [comment]
damn

**Comment by elthehablo** [comment]
Awesome work @sudlud . I'll give this a test and a review promptly

File: src/server/scripts/Events/firework_show/firework_show.cpp [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    go_firework_show(GameObject* go) : GameObjectAI(go)
+    {
+        _curIdx = 0;
+        _showRunning = false;
+        _show = nullptr;
+
+        initShow();
+    }
+
+    void initShow()
```
Comment (elthehablo) [inline src/server/scripts/Events/firework_show/firework_show.cpp:57]: We use Pascal Case for function names

**Review by elthehablo — CHANGES_REQUESTED** [review]

**Comment by elthehablo** [comment]
Works well, nice work

**Review by elthehablo — APPROVED** [review]

File: src/server/scripts/Events/firework_show/firework_show.cpp (hunk 1 above)
Comment (sudlud) [reply src/server/scripts/Events/firework_show/firework_show.cpp:57]: I'll update the file later today

**Pull request merged**